### PR TITLE
conv_pointwise_type parameter 

### PR DESCRIPTION
**Pull request by bmwshop**
# What does this PR do ?

Adding the conv_pointwise_type parameter that to ConformerEncoder enabling the use of nn.Linear() for a 10% acceleration 

**Collection**: ASR

# Changelog 
Adding the conv_pointwise_type parameter to 
- nemo/collections/asr/modules/conformer_encoder.py : adding the conv_pointwise_type parameter and propagating it to ConformerLayer
- nemo/collections/asr/parts/submodules/conformer_modules.py: adding conv_pointwise_type parameter to ConformerLayer and propagating it to ConformerConvolution.
- in ConformerConvolution, if conv_pointwise_type is 'linear', use nn.Linear() instead of nn.Conv1d()

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
+model.encoder.conv_pointwise_type='linear'
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/asr/parts/submodules/conformer_modules.py [hunk 1]
```diff
@@ -351,13 +370,25 @@ def __init__(
...
+        else:
+            raise ValueError(
+                f"'{conv_pointwise_type}' is not not a valid value for 'conv_pointwise_type', "
+                f"valid values can be from ['conv1d', 'linear']"
+            )
 
     def forward(self, x, pad_mask=None, cache=None, cache_next=None):
-        x = x.transpose(1, 2)
-        x = self.pointwise_conv1(x)
+        if self.conv_pointwise_type == 'conv1d':
```
Comment (titu1994) [inline nemo/collections/asr/parts/submodules/conformer_modules.py:386]: There weren't transposes before for conv1d, why are they there now ?

File: nemo/collections/asr/parts/submodules/conformer_modules.py [hunk 2]
```diff
@@ -381,8 +412,12 @@ def forward(self, x, pad_mask=None, cache=None, cache_next=None):
             x = self.batch_norm(x)
 
         x = self.activation(x)
-        x = self.pointwise_conv2(x)
-        x = x.transpose(1, 2)
+        if self.conv_pointwise_type == 'conv1d':
```
Comment (titu1994) [inline nemo/collections/asr/parts/submodules/conformer_modules.py:415]: Same as above

File: nemo/collections/asr/parts/submodules/conformer_modules.py (hunk 1 above)
Comment (bmwshop) [reply nemo/collections/asr/parts/submodules/conformer_modules.py:386]: if conv_pointwise_type == 'conv1d', there is no change. If conv_pointwise_type == 'linear', then we transpose..

**Review by titu1994 — COMMENTED** [review]
Looks good. But needs to be documented in Nemo ASR docs config section

**Comment by bmwshop** [comment]
> Looks good. But needs to be documented in Nemo ASR docs config section

should we add to this PR or do a separate PR for docs?

**Comment by titu1994** [comment]
Same PR would be good

**Comment by bmwshop** [comment]
> Same PR would be good
Added some docs - was not sure how to best do it, ended up appending a section on conformer encoder

File: docs/source/asr/configs.rst [hunk 3]
```diff
@@ -657,6 +657,68 @@ The transducer model is comprised of three models combined. One of these models
 
 The only condition that needs to be met is that **the final layer of the acoustic model must have the hidden dimension defined in ``model_defaults.enc_hidden``**.
 
+Conformer Encoder
```
Comment (titu1994) [inline docs/source/asr/configs.rst:660]: Is this for fast conformer ? Some flags are for fast conformer

File: nemo/collections/asr/parts/submodules/conformer_modules.py [hunk 4]
```diff
@@ -351,13 +370,25 @@ def __init__(
...
 
         self.activation = Swish()
-        self.pointwise_conv2 = nn.Conv1d(
-            in_channels=dw_conv_input_dim, out_channels=d_model, kernel_size=1, stride=1, padding=0, bias=True
-        )
+        if self.conv_pointwise_type == 'conv1d':
+            self.pointwise_conv2 = nn.Conv1d(
+                in_channels=dw_conv_input_dim, out_channels=d_model, kernel_size=1, stride=1, padding=0, bias=True
+            )
+        elif self.conv_pointwise_type == 'linear':
```
Comment (titu1994) [inline nemo/collections/asr/parts/submodules/conformer_modules.py:377]: I wonder. Conv1d with kernel 1 is identical to linear layer, so what if we do something like silently convert the conv kernel shape to linear compatible weights and don't use explicitly the Lineae layer but use functional.linear() instead ? 

My assumption is we don't want to break checkpoint compatibility - and we don't need to. We can take a model trained on conv1d and use the weight in a F.linear() after weight conversion - this way no checkpoint incompatibility is there for old models but still there is speedup

**Review by titu1994 — COMMENTED** [review]
Minor comments

**Comment by ArtyomZemlyak** [comment]
@bmwshop 
Hi!
I would like to ask why you decided to close the pr?

**Pull request closed without merging**